tests/std/tests/GH_004930_char_traits_user_specialization/test.cpp

```diff
@@ -0,0 +1,210 @@
+    }
+
+    static constexpr bool lt(const char_type c, const char_type d) noexcept {
+        return (static_cast<unsigned char>(c) & odd_mask) < (static_cast<unsigned char>(d) & odd_mask);
+    }
+
+    static constexpr int compare(const char_type* const c, const char_type* const d, const size_t n) noexcept {
+        for (size_t i = 0; i != n; ++i) {
+            int ci = static_cast<unsigned char>(c[i]) & odd_mask;
+            int di = static_cast<unsigned char>(d[i]) & odd_mask;
+            int r  = ci - di;
+            if (r != 0) {
+                return r;
+            }
+        }
+
+        return 0;
+    }
+
+    static constexpr size_t length(const char_type* const p) noexcept {
+        const char_type* c = p;
+        while (static_cast<unsigned char>(*c) != 0) {
+            ++c;
+        }
+
+        return static_cast<size_t>(c - p);
+    }
+
+    static constexpr const char_type* find(const char_type* p, const size_t n, const char_type c) noexcept {
+        for (size_t i = 0; i != n; ++i) {
+            if (eq(p[i], c)) {
+                return p + i;
+            }
+        }
+
+        return nullptr;
+    }
+
+    static CONSTEXPR20 char_type* move(char_type* s, const char_type* p, const size_t n) noexcept {
+#if _HAS_CXX20
+        if (is_constant_evaluated()) {
+            bool is_dst_in_src_range = false;
+            for (size_t i = 0; i != n; ++i) {
+                if (p + i == s) {
+                    is_dst_in_src_range = true;
+                    break;
+                }
+            }
+
+            if (is_dst_in_src_range) {
+                for (size_t i = n; i != 0;) {
+                    --i;
+                    s[i] = p[i];
+                }
+            } else {
+                for (size_t i = 0; i != n; ++i) {
+                    s[i] = p[i];
+                }
+            }
+        } else
+#endif // _HAS_CXX20
+        {
+            memmove(s, p, n);
+        }
+        return s;
+    }
+
+    static CONSTEXPR20 char_type* copy(char_type* s, const char_type* p, const size_t n) noexcept {
+#if _HAS_CXX20
+        if (is_constant_evaluated()) {
+            for (size_t i = 0; i != n; ++i) {
+                s[i] = p[i];
+            }
+        } else
+#endif // _HAS_CXX20
+        {
+            memmove(s, p, n);
+        }
+        return s;
+    }
+
+    static constexpr void assign(char_type& r, const char_type& d) noexcept {
+        r = d;
+    }
+    static CONSTEXPR20 char_type* assign(char_type* const s, const size_t n, const char_type c) {
+#if _HAS_CXX20
+        if (is_constant_evaluated()) {
+            for (size_t i = 0; i != n; ++i) {
+                s[i] = c;
+            }
+        } else
+#endif // _HAS_CXX20
+        {
+            memset(s, static_cast<unsigned char>(c), n);
+        }
+        return s;
+    }
+
+    static constexpr bool not_eof(int) noexcept {
+        return true;

```

Comment:
I recognize you copied my repro and it didn't implement `not_eof` right.
Maybe worth attempting, as it might be used in streaming tests as well someday.

Reply:
Fixed. I found that `length` also needed to be fixed.